### ISSUE TITLE
Proposal: Add a function to return a partitioned collection of coded locations

### ISSUE THREAD

**Issue by DrCuriosity** [GNS-Science/nzshm-common-py#40]
For use cases where an iterable collection of coded locations need to be partitioned, e.g.:

## (From Toshi Hazard Store)
```py
# get the 0.1 degree grid locations for all the locations
partition_codes = sorted(set([CodedLocation(lat=loc[0], lon=loc[1], resolution=0.1).code for loc in location_list]))
for partition_code in partition_codes:
    ... etc
```

It could be useful to create a data structure that collects the locations into their respective partitions in one go.

Given a `dict[Iterable]` data structure like:

- `-41.3~174.6`: LocationPartition(lat=-41.3, lon=174.6, resolution=0.1)
    - CodedLocation(lat=-41.31, long=174.61, resolution=0.01)
    - CodedLocation(lat=-41.32, long=174.61, resolution=0.01)
    - &hellip;
- `-41.4~174.6`: LocationPartition(lat=-41.4, lon=174.6, resolution=0.1)
    - CodedLocation(lat=-41.41, long=174.61, resolution=0.01)
    - CodedLocation(lat=-41.42, long=174.61, resolution=0.01)
    - &hellip;
- &hellip;

This could then be processed with:

```py
for partition in partition_locations(locations, resolution=0.1):
    for location in partition.locations:
        # Some calculation referencing the CodedLocation here

    # Any aggregate calculations for the partition

# Any aggregate calculations across all partitions
```

**Comment by DrCuriosity** [comment]
We might have to be careful to document the rounding procedures here, so that it is well understood which points will be collected into which partitions.

**Comment by chrisbc** [comment]
> We might have to be careful to document the rounding procedures here, so that it is well understood which points will be collected into which partitions.

absolutely, it's a critical part of the whole location coding scheme, but I this discussion belongs more with `CodedLocation` itself.

**Comment by chrisbc** [comment]
@DrCuriosity I wanna sleep on  using the term 'partition' here. that name comes from parquet / arrow world. We could just as easily use  location_bin, location_set, location_group. Maybe it'd fine ?? @chrisdicaprio  thoughts?

**Comment by DrCuriosity** [comment]
> absolutely, it's a critical part of the whole location coding scheme, but I this discussion belongs more with `CodedLocation` itself.

Agreed. We can probably just add a Markdown link here that references the logic there, for convenience.

**Comment by chrisdicaprio** [comment]
`location_bin` is the preferred term (my vote)
